Add CIDR IP address range hash to fingerprint metadata

eduardoumpierre opened the pull request:
#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

Add CIDR IP address range hash to fingerprint metadata.

This additional data will be used on the card testing mitigation mechanism on server.

Related to 2340-gh-automattic/woocommerce-payments-server.

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* All tests are passing

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

eduardoumpierre committed:
Add 3 octet IP hash to fingerprint metadata

This additional data will be used on the card testing
mitigation mechanism.

eduardoumpierre committed:
Add changelog entry

eduardoumpierre committed:
Update 3 octet IP hash key to match Stripe doc

Stripe limits the key to be up to 40 characters long

kalessil commented on includes/fraud-prevention/class-buyer-fingerprinting-service.php:
```diff
@@ -54,15 +54,33 @@ public function hash_data_for_fraud_prevention( string $data ): string {
+	 * @return string 3 octet IP address (192.192.192).
+	 */
+	public function get_3_octet_from_ip( string $ip_address ): string {
+		$ip_address_octets = explode( '.', $ip_address );
```
Is there a specific reason why we are not taking into account IPv6 here?

eduardoumpierre replied:
Thanks for your comment @kalessil. In fact, we didn't consider IPv6 while thinking about this solution. I'll adapt the nomenclatures and the code to be able to handle both versions.

eduardoumpierre replied:
Updated on [084a732](https://github.com/Automattic/woocommerce-payments/pull/4709/commits/084a7328577a9070bf78680127a3644d27d3e385) and [1d3e464](https://github.com/Automattic/woocommerce-payments/pull/4709/commits/1d3e4647ffaa5d864fda90aba3e214604605f4a0).

eduardoumpierre committed:
Add CIDR_Range class to handle CIDR ranges for IPv4 and IPv6 addresses

eduardoumpierre committed:
Update fingerprint metadata to use IP range instead of IPv4 octets

Fraud prevention metadata changed from fraud_prevention_data_shopper_3_octets_hash
to fraud_prevention_data_shopper_range_hash.

eduardoumpierre committed:
Remove unused 3 octet IP helper function

eduardoumpierre committed:
Update changelog entry

tpaksu commented on includes/class-cidr-range.php:
```diff
@@ -0,0 +1,139 @@
+	public static function from_ipv6( string $ip, int $prefix = self::IPV6_PREFIX ) {
+		try {
+			// Split in address and prefix length.
+			list( $address_string, $prefix_length ) = explode( '/', "$ip/$prefix" );
```
Why did we do this? To eliminate the $prefix parameter if the address already contains the prefix? Otherwise, won't be `$address_string === $ip` and `$prefix_length === $prefix`?

eduardoumpierre replied:
Thanks for catching this, those lines went completely unnoticed. Fixed on [8dc1dec](https://github.com/Automattic/woocommerce-payments/pull/4709/commits/8dc1decfc021877a584bd04310b8d9c6e6c2c3ff).

tpaksu commented on includes/class-cidr-range.php:
```diff
@@ -0,0 +1,139 @@
+	 */
+	public static function from_ipv4( string $ip, int $prefix = self::IPV4_PREFIX ) {
+		try {
+			list( $ip, $mask ) = explode( '/', "$ip/$prefix" );
```
The same question also applies to this.

eduardoumpierre replied:
Fixed on [8dc1dec](https://github.com/Automattic/woocommerce-payments/pull/4709/commits/8dc1decfc021877a584bd04310b8d9c6e6c2c3ff).

tpaksu reviewed (approved):
- Code is understandable (except the IPv6 range calculation, I got a bit lost on the bits loop). 
- Newly introduced variables and functions are named good, 
- Testing instructions are missing how to test it, but I could test it with going to my UPE-enabled checkout page with some items in the cart, and see if the payment intent creation event contains the fingerprint. It exists on the payment intent.
<img width="739" alt="image" src="https://user-images.githubusercontent.com/3295/190570613-35dba1c7-f87b-4472-979d-9d28bae33c80.png">

All look good, I have one question that I wonder why it's used that way, which is mostly a nitpick, and won't change the approach here. Feel free to skip it. :shipit:

eduardoumpierre committed:
Remove unnecessary string splitting on CIDR Range

All the parameters are already available in the methods’ scope and
splitting the string isn't necessary.

eduardoumpierre commented:
Closing this pull request due to a team decision based on pMz3w-fWH-p2#comment-101923 and paJDYF-5Dc-p2, we will work on a more trusted fingerprint implementation in the next project phase.
